pkg/crypto/certloader/watcher.go

```diff
@@ -118,61 +125,59 @@ func (w *Watcher) Watch() <-chan struct{} {
 	// prepare the ready channel to be returned. We will close it exactly once.
 	var once sync.Once
 	ready := make(chan struct{})
-
-	// build maps for the CA files and keypair files to help detecting what has
-	// changed in order to reload only the appropriate certificates.
-	keypairMap := make(map[string]struct{})
-	caMap := make(map[string]struct{})
-	if w.FileReloader.certFile != "" {
-		keypairMap[w.FileReloader.certFile] = struct{}{}
-	}
-	if w.FileReloader.privkeyFile != "" {
-		keypairMap[w.FileReloader.privkeyFile] = struct{}{}
-	}
-	for _, path := range w.FileReloader.caFiles {
-		caMap[path] = struct{}{}
-	}
+	var reload <-chan time.Time
 
 	go func() {
 		defer w.fswatcher.Close()
 		for {
 			select {
 			case event := <-w.fswatcher.Events:
-				path := event.Name
-				log := w.log.WithFields(logrus.Fields{
-					logfields.Path: path,
+				w.log.WithFields(logrus.Fields{
+					logfields.Path: event.Name,
 					"operation":    event.Op,
-				})
-				log.Debug("Received fsnotify event")
-				_, keypairUpdated := keypairMap[path]
-				_, caUpdated := caMap[path]
-				createOperation := event.Op&fsnotify.Create == fsnotify.Create
-				if (keypairUpdated || caUpdated) && createOperation {
-					// A file we are interested in has been created. We might
-					// have been watching only over its directory until now,
-					// add it to the watcher since it exists.
-					if err := w.fswatcher.Add(path); err != nil {
-						log.WithError(err).Warnf("Failed to add %s to fsnotify watcher", logfields.Path)
-					}
+				}).Debug("Received fsnotify event")
+
+				// Kubernetes implements volume mounts of Secrets and ConfigMaps
+				// either by creating a symlink to the certificate/keyfile, or
+				// by creating a symlink to a folder containing the files
+				// (for projected volume mounts with subpaths). This allows it
+				// to update all symlink targets atomically.
+				// Unfortunately, this makes also makes it hard to derive which
+				// file has been updated, as event.Name alone does not provide
+				// enough context. Instead, we would have to to re-resolve
+				// the symlinks each time and and update the watched files to
+				// contain the new targets w.fswatcher.
+				//
+				// Instead of doing that, we implement a simpler and hopefully
+				// more robust approach of simply watching the volume mount
+				// directory (see findKubernetesVolumeMount in newFsWatcher).
+				// If there is any update on the volume mount, we coascale all
+				// events that we obtain within a certain time window and
+				// then reload _all_ certificates and keypairs at once.
+				// This drastically simplifies the implementation. As an added
+				// bonus, this approach also deals with the non-Kubernetes case
+				// where the keypair and ca-certificates files are updated
+				// non-atomically in individual writes (which this logic
+				// also coalesces).
+				if reload == nil {
+					reload = time.After(watcherCoalesceWindow)
 				}
```

Comment:
am I reading this correctly? doesn't this mean that this is essentially polling?

Reply:
Not really. The current approach suffers from spurious unrelated events causing unnecessary reloads, but it's still very different from polling. Note that the reload is only triggered when there is an event, not on a timer (which is what polling would do).

But I put the PR back into draft mode, there might be a more precise way to deal with the K8s sequence.

Reply:
Resolved as outdated. The new approach does not suffer from spurious events anymore.